libs/wingc/src/fold.rs

```diff
@@ -0,0 +1,438 @@
+use crate::ast::{
+	ArgList, CatchBlock, Class, ClassField, Constructor, ElifBlock, Expr, ExprKind, FunctionBody, FunctionDefinition,
+	FunctionParameter, FunctionSignature, FunctionTypeAnnotation, Interface, InterpolatedString, InterpolatedStringPart,
+	Literal, Reference, Scope, Stmt, StmtKind, Symbol, TypeAnnotation, UserDefinedType,
+};
+
+/// Similar to the `visit` module in `wingc` except each method takes ownership of an
+/// AST node instance of a reference to it, and returns a new AST node instance.
+/// This trait can be useful for AST transformations.
+pub trait Fold {
```

Comment:
"Fold" is not a term I have heard for this, but I see it's a rust thing and I can't think of a better name 😅 

Maybe ASTTransformer or something, but I guess that kinda implies it's also changing the type (which this isn't).

Reply:
yeah, the use of the word "fold" actually goes pretty far back as a piece of functional programming jargon: https://en.wikipedia.org/wiki/Fold_(higher-order_function)

Besides being historically neat, the fact that "fold" is only four characters will save us a lot of keystrokes here, and it should be OK since it's compiler internal. I can probably improve the description here

Reply:
+1 on the name

Reply:
When I said it's a rust thing I moreso meant that it's described here as being used in this pattern and being used in the rust compiler https://rust-unofficial.github.io/patterns/patterns/creational/fold.html, my bad.

I hadn't heard the term fold but I'm familiar with reducers. This is the first time I've seen a reducer that doesn't actually "reduce" anything, so it looked more like a map. From reading more I think I understand the distinction, but this was definitely not obvious to me on the name alone. I'll just take that as a personal issue of something I should understand.